Gap-5: App Service Operate (B+ → A) — SKU selection, custom domains, networking (#1639)
microsoft/GitHub-Copilot-for-Azure: plugin/skills/azure-prepare/references/services/app-service/networking.md

# App Service Networking

VNet integration, Private Endpoints, Access Restrictions, and Hybrid Connections.

## Feature Availability by SKU

| Feature | Free/Basic | Standard | Premium | Isolated |
|---------|:-:|:-:|:-:|:-:|
| VNet integration (outbound) | ❌ | ✅ | ✅ | ✅ (native) |
| Private Endpoints (inbound) | ❌ | ❌ | ✅ | ✅ |
| Access Restrictions | ✅ | ✅ | ✅ | ✅ |
| Hybrid Connections | ❌ | 25 | 200 | 200 |
| Service Endpoints | ❌ | ✅ | ✅ | ✅ |

> 

Reviewer: +1

> 

Reviewer 2: Updated table to resolve comment

## VNet Integration (Outbound)

Routes outbound traffic from the app through a VNet subnet, enabling access to private resources (databases, storage, VMs).

### Subnet Requirements

| Requirement | Value |
|------------|-------|
| Minimum subnet size | `/26` (64 addresses) recommended |
| Delegation | `Microsoft.Web/serverFarms` |
| Dedicated | One subnet per App Service plan |

### Bicep — VNet Integration

```bicep
resource subnet 'Microsoft.Network/virtualNetworks/subnets@2023-11-01' = {
  parent: vnet
  name: 'app-service-subnet'
  properties: {
    addressPrefix: '10.0.1.0/26'
    delegations: [
      {
        name: 'Microsoft.Web.serverFarms'
        properties: { serviceName: 'Microsoft.Web/serverFarms' }
      }
    ]
  }
}

resource webApp 'Microsoft.Web/sites@2022-09-01' = {
  name: appName
  location: location
  properties: {
    serverFarmId: appServicePlan.id
    virtualNetworkSubnetId: subnet.id

>     vnetRouteAllEnabled: true // route all outbound through VNet

Reviewer 2: vnetRouteAllEnabled is a legacy setting and customers should use the outboundVnetRouting.allTraffic property to enable routing of all traffic.  Using this property also allows customers to audit the behaviour with a built-in policy.  Reference - https://learn.microsoft.com/azure/app-service/configure-vnet-integration-routing#configure-application-routing

>     vnetRouteAllEnabled: true // route all outbound through VNet

Reviewer 2: ```suggestion
    outboundVnetRouting.allTraffic: true // route all outbound through VNet
```

>   }

Reviewer 2: @copilot update api version to 2024-11-01

}
```

> > 💡 **Tip:** Set `vnetRouteAllEnabled: true` to route ALL outbound traffic through the VNet. Without this, only RFC1918 traffic is routed through the VNet.

Reviewer 2: @copilot refresh this suggestion

## Private Endpoints (Inbound)

Expose the app on a private IP address within your VNet. Public access can be disabled entirely.

### Bicep — Private Endpoint

```bicep
resource privateEndpoint 'Microsoft.Network/privateEndpoints@2023-11-01' = {
  name: '${appName}-pe'
  location: location
  properties: {
    subnet: { id: privateEndpointSubnet.id }
    privateLinkServiceConnections: [
      {
        name: '${appName}-connection'
        properties: {
          privateLinkServiceId: webApp.id

>           groupIds: ['sites']

Reviewer 2: @copilot please re-check

        }
      }
    ]
  }

> }

Reviewer 2: @copilot please re-check

resource privateDnsZone 'Microsoft.Network/privateDnsZones@2020-06-01' = {
  name: 'privatelink.azurewebsites.net'
  location: 'global'
}

resource dnsLink 'Microsoft.Network/privateDnsZones/virtualNetworkLinks@2020-06-01' = {
  parent: privateDnsZone
  name: '${vnet.name}-link'
  location: 'global'
  properties: {
    virtualNetwork: { id: vnet.id }
    registrationEnabled: false
  }

> }

Reviewer 2: @copilot please re-check

```

> ⚠️ **Warning:** Private Endpoints require Premium (P1v3+) or Isolated tier. The private DNS zone `privatelink.azurewebsites.net` must be linked to the VNet for name resolution.

## Access Restrictions

Control inbound access with IP-based or service-tag rules. Available on all SKUs.

### Bicep — Access Restrictions

```bicep
siteConfig: {
  ipSecurityRestrictions: [
    {
      name: 'allow-office'
      priority: 100
      action: 'Allow'
      ipAddress: '203.0.113.0/24'
    }
    {
      name: 'deny-all'
      priority: 2147483647
      action: 'Deny'

>       ipAddress: 'Any'

Reviewer 2: Updated on line 128

    }
  ]
  scmIpSecurityRestrictionsUseMain: true
}
```

> 💡 **Tip:** Always restrict the SCM/Kudu site too. Use `scmIpSecurityRestrictionsUseMain: true` to inherit main site rules, or define separate SCM rules.

## Hybrid Connections

Connect to on-premises resources without VPN. Requires Standard tier or higher. Uses Hybrid Connection Manager (HCM) agent on-premises relaying through Azure Relay.

> ⚠️ **Warning:** Each Hybrid Connection maps to a single host:port endpoint. Standard tier supports 25; Premium/Isolated support 200.

## Troubleshooting

| Issue | Cause | Fix |
|-------|-------|-----|

> | Cannot reach private DB | VNet integration not enabled | Enable VNet integration; check `vnetRouteAllEnabled` |

Reviewer 2: @copilot update your suggested change to check for the full property - outboundVnetRouting.allTraffic

| DNS resolution fails | Private DNS zone not linked | Link `privatelink.*` DNS zone to VNet |
| Access restriction not working | Priority ordering wrong | Lower numbers = higher priority; check rule order |
| Hybrid Connection timeout | HCM not running | Verify HCM service status on-premises |

> | Outbound traffic blocked | NSG rules on subnet | Allow outbound to required services in NSG |

Reviewer 2: Have added CLI examples

      allTraffic: true // route all outbound through VNet
    }
  }
}
```

> 💡 **Tip:** Set `outboundVnetRouting.allTraffic: true` to route ALL outbound traffic through the VNet. Without this, only RFC1918 traffic is routed through the VNet.

## Private Endpoints (Inbound)

Expose the app on a private IP address within your VNet. Public access can be disabled entirely.

### Bicep — Private Endpoint

```bicep
resource privateEndpoint 'Microsoft.Network/privateEndpoints@2023-11-01' = {
  name: '${appName}-pe'
  location: location
  properties: {
    subnet: { id: privateEndpointSubnet.id }
    privateLinkServiceConnections: [
      {
        name: '${appName}-connection'
        properties: {
          privateLinkServiceId: webApp.id
          groupIds: ['sites']
        }
      }
    ]
  }
}

resource privateDnsZone 'Microsoft.Network/privateDnsZones@2020-06-01' = {
  name: 'privatelink.azurewebsites.net'
  location: 'global'
}

resource dnsLink 'Microsoft.Network/privateDnsZones/virtualNetworkLinks@2020-06-01' = {
  parent: privateDnsZone
  name: '${vnet.name}-link'
  location: 'global'
  properties: {
    virtualNetwork: { id: vnet.id }
    registrationEnabled: false
  }
}
```

> > ⚠️ **Warning:** Private Endpoints require Premium (P1v3+) or Isolated tier. The private DNS zone `privatelink.azurewebsites.net` must be linked to the VNet for name resolution.

Reviewer 2: ```suggestion
> ⚠️ **Warning:** Private Endpoints require Basic (B1+) or higher tier. The private DNS zone `privatelink.azurewebsites.net` must be linked to the VNet for name resolution.
```

## Access Restrictions

Control inbound access with IP-based or service-tag rules. Available on all SKUs.

### Bicep — Access Restrictions

```bicep
siteConfig: {
  ipSecurityRestrictions: [
    {
      name: 'allow-office'
      priority: 100
      action: 'Allow'
      ipAddress: '203.0.113.0/24'
    }
    {
      name: 'deny-all'
      priority: 2147483647
      action: 'Deny'
      ipAddress: 'Any'
    }
  ]
  scmIpSecurityRestrictionsUseMain: true
}
```

> 💡 **Tip:** Always restrict the SCM/Kudu site too. Use `scmIpSecurityRestrictionsUseMain: true` to inherit main site rules, or define separate SCM rules.

## Hybrid Connections

> Connect to on-premises resources without VPN. Requires Standard tier or higher. Uses Hybrid Connection Manager (HCM) agent on-premises relaying through Azure Relay.

Reviewer 2: ```suggestion
Connect to on-premises resources without VPN. Requires Basic tier or higher. Uses Hybrid Connection Manager (HCM) agent on-premises relaying through Azure Relay.
```

# App Service Networking

VNet integration, Private Endpoints, Access Restrictions, and Hybrid Connections.

## Feature Availability by SKU

> | Feature | Free/Basic | Standard | Premium | Isolated |

Reviewer 2: ```suggestion
| Feature | Free | Basic | Standard | Premium | Isolated |
```

> |---------|:-:|:-:|:-:|:-:|

Reviewer 2: ```suggestion
|---------|:-:|:-:|:-:|:-:|:-:|
```

> | VNet integration (outbound) | ❌ | ✅ | ✅ | ✅ (native) |

Reviewer 2: ```suggestion
| VNet integration (outbound) | ❌ | ✅ | ✅ | ✅ | ✅ (native) |
```

> | Private Endpoints (inbound) | ❌ | ❌ | ✅ | ✅ |

Reviewer 2: ```suggestion
| Private Endpoints (inbound) | ❌ | ✅ | ✅ | ✅ | ✅ |
```

> | Access Restrictions | ✅ | ✅ | ✅ | ✅ |

Reviewer 2: ```suggestion
| Access Restrictions | ✅ | ✅ | ✅ | ✅ | ✅ |
```

> | Hybrid Connections | ❌ | 25 | 200 | 200 |

Reviewer 2: ```suggestion
| Hybrid Connections | ❌ | 5 | 25 | 200 | 200 |
```

> | Access to service-endpoint-protected resources | ❌ | ✅ | ✅ | ✅ |

Reviewer 2: ```suggestion
| Access to service-endpoint-protected resources | ❌ | ✅ | ✅ | ✅ | ✅ |
```

# App Service Networking

VNet integration, Private Endpoints, Access Restrictions, and Hybrid Connections.

## Feature Availability by SKU

| Feature | Free | Basic | Standard | Premium | Isolated |
|---------|:-:|:-:|:-:|:-:|:-:|
| VNet integration (outbound) | ❌ | ✅ | ✅ | ✅ | ✅ (native) |
| Private Endpoints (inbound) | ❌ | ✅ | ✅ | ✅ | ✅ |
| Access Restrictions | ✅ | ✅ | ✅ | ✅ | ✅ |

> | Hybrid Connections | ❌ | 5 | 25 | 200 | 200 |

Reviewer 2: @copilot your finding is incorrect, all table rows in this doc start with single pipe
